src/server/common_services/auth_server.js

```diff
@@ -660,12 +662,45 @@ function _get_auth_info(account, system, authorized_by, role, extra) {
     return response;
 }
 
+// has_bucket_permission returns true if either the account is the owner of the bucket
+// or the account has explicit permission to access the bucket via the Bucket Policy.
+//
+// NOTE: If the owner account has is denied access as per the policy then the owner
+// also is denied access to the bucket.
 function has_bucket_permission(bucket, account) {
-    return _.get(account, 'allowed_buckets.full_permission', false) ||
-        _.find(
-            _.get(account, 'allowed_buckets.permission_list') || [],
-            allowed_bucket => String(allowed_bucket._id) === String(bucket._id)
-        );
+    dbg.log0('has_bucket_permission:', bucket.name, account.email, bucket.owner_account.email);
+    const is_owner = bucket.owner_account.email.unwrap() === account.email.unwrap();
+    const bucket_policy = bucket.s3_policy;
+
+    if (!bucket_policy) {
+        return is_owner;
+    }
+
+    const bucket_policy_evaluation = _.flatten(
```

Comment:
I would add an explanation about what we are doing here. maybe we should check for an * or pass the specfic action which I think will be able to understand. so maybe we should just check that he has full control on the bucket.

Reply:
I also wanted to perform check only for the requested action however, when I tried to see the call chain for the function, it seemed that the original action gets lost (for example when getting called from the `bucket_server.js`) which makes it difficult to pass it as one of the parameters to the function.

This method just checks if there is any policy which can be used to allow user the access to the bucket (to try to replicate `permission_list` behaviour)

Reply:
ok, so It make sense. but let's add a comment explaining that. I think usually what we wanted to see is if the user suppose to see this bucket, so maybe we can just check if he can read the bucket/list it objects. WDYT?

Reply:
Oh yes, that's what I tried in the first iteration. Turns out that functions like `load_bucket` (which indirectly relies on this function) is invoked in multiple operations in `object_server.js` including in `create_object_upload` etc.

Taking the example of `create_object_upload`, if a user does have upload permission (regardless of the list and get permissions), the operation should succeed right?

Assuming the above assumption is correct, the `has_bucket_permission` checks if a user can do _something_ with the bucket or not.

---
I also explored if we can get rid of permission checks in the utility functions `load_bucket`, `find_bucket`. They seemed redundant assuming that the request has already flown through the initial `authorization` checks (where we do have the information of "action" as well). However, what I realized and I might be wrong here and please correct if I am - We can invoke RPCs to these entities which will bypass the REST flow essentially bypassing the REST authorization flow, hence I left the permission checks in these methods.

Reply:
I agree that it seems that authorization needs to happen only in its own part, so maybe our goal should be to remove this checks from both load_bucket and find_bucket, and leave it only to s3_rest authorization. please check if you find any place where you can see such behavior where we authorize the request regarding one bucket and then tries to access a different one (other than the triggers that we already discussed and needs to be removed in the future)

Reply:
You were right @jackyalbo. We should be able to get rid of authorization inside `load_bucket`, `find_bucket` and `list_buckets` because we the request gets already authorized in the s3_rest flow and we never request for buckets which were not part of original user request.

A caveat though: I might be wrong here so please correct me if I am - in our tests we often make direct RPC calls and RPC calls only go through authentication (in their middlewares), so won't the tests will fail? Also as far as I understand, RPC calls can be made independent of operator, in that case it might be problematic to skip authorization right?

Reply:
I still think this can go as well... for sure the bucket policy part as we didn't check it before this change as well, so probably not need for it now all of a sudden. I think commands from RPC should be handled on it's own. But If anything jumps to your eyes that some call suppose to be blocked but doesn't - let's discuss.